### ISSUE TITLE
Installation on OSX

### ISSUE THREAD

**Issue by kayceesrk** [ocamllabs/ocaml-multicore#57]
Fails with 

```
ld: illegal thread local variable reference to regular symbol _caml_domain_state
```

See #56.

**Comment by bluddy** [comment]
A google search leads to [this](https://github.com/gambit/gambit/issues/109) similar issue with Gambit, which is a language implementation as well. The problem is that OSX doesn't allow dynamic linking to an external thread-local variable. Their solution is to have different implementations of thread local state abstracted away behind a macro, as seen [here](https://github.com/gambit/gambit/blob/master/lib/os_thread.c#L351) -- they either return the extern tls pointer or they use pthread's thread local storage which is supposed to be slower.

Here is some more information from our linking process, invoking the compilation line that causes the problem, while emitting the compilation commands:

```
../../boot/ocamlrun ../../ocamlopt -verbose -nostdlib -I ../../stdlib -shared -o unix.cmxs -I . unix.cmxa
+ clang -arch x86_64 -c -o 'unix.cmxs.startup.o' '/var/folders/f_/0r8gp7l97kdgx39vgzc49v9h0000gn/T/camlstartupabb3a0.s'
+ gcc -bundle -flat_namespace -undefined suppress -Wl,-no_compact_unwind -o 'unix.cmxs'   '-L../../stdlib' '-L.'  'unix.cmxs.startup.o' 'unix.a' '-lunix'
ld: illegal thread local variable reference to regular symbol _caml_domain_state for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
File "caml_startup", line 1:
Error: Error during linking
```

Since we seem to require building a dynamically linked library here (unix.cmxs) which accesses TLS, we probably need to use the strategy I linked to.

**Comment by kayceesrk** [comment]
Thanks @bluddy. The solution with the macro seems very reasonable. Would you like to take a shot at implementing the fix? Eventually, the goal is to move all thread-local variable to domain_local state at the end of minor heap area [0]. But in the intermediate time, it is probably worth making sure we can compile programs. 

[0] https://github.com/ocamllabs/ocaml-multicore/blob/master/byterun/caml/domain_state.tbl

**Comment by bluddy** [comment]
I see. So there are currently 2 TLS pointers, domain_self and caml_domain_state, and you're planning on eventually having just domain_self? Is that right?

**Comment by bluddy** [comment]
Does the bytecode use this domain_state machinery?

**Comment by jberdine** [comment]
I'm hitting this and wondering if there has been any progress?

**Comment by kayceesrk** [comment]
See PR https://github.com/ocamllabs/ocaml-multicore/pull/65.

**Comment by kayceesrk** [comment]
Fixed by #103. Thanks @bluddy for the PR.
